### ISSUE TITLE
Machine-Checkable Formal Semantics

### ISSUE THREAD

**Issue by lorenzleutgeb** [WebAssembly/design#1198]
TL;DR: What's the state of a formal, machine-checkable semantics for wasm?

Hello!

This is meant as a question and basis for discussion.

WebAssembly is an attractive compilation target (golang/go#18892, [rust-lang-nursery/rust-wasm](https://github.com/rust-lang-nursery/rust-wasm), [emscripten](https://kripken.github.io/emscripten-site/index.html), ...). And it would be nice to have a spec that is as precise as possible, to eliminate ambiguities between the languages that compile to wasm. From what I see, you, the designers, are very aware of these facts and steering towards that goal.

However, my understanding of a "formal" specification is more rigid. Instead of a carefully crafted specification in natural language, I am interested in a formulation using mathematical language. Also, I would want it to be machine-readable and machine-checkable.

Example of such specification efforts:
 * [JsCert](https://github.com/jscert/jscert) (paper [as PDF](http://www.doc.ic.ac.uk/~maffeis/papers/popl14.pdf) and [via DOI](http://doi.acm.org/10.1145/2535838.2535876))
 * A Formal Semantics for C++ (paper [as PDF](https://ts.data61.csiro.au/publications/nicta_full_text/1203.pdf) and [at CSIRO](https://publications.csiro.au/rpr/pub?pid=nicta:1203)) by @mn200
 * [RustBelt](https://plv.mpi-sws.org/rustbelt/) ([paper via DOI](http://doi.org/10.1145/3158154))
 * [The Semantics of CakeML](https://github.com/CakeML/cakeml/tree/master/semantics)
 * [The Definition of Standard ML](https://mitpress.mit.edu/books/definition-standard-ml)

As the authors of JsCert note in the paper referenced above:

> Our challenge is to convince ourselves and others that JSCert is indeed an accurate formulation of the ES5 English specification. We designed JSCert to follow the structure of the ES5 English standard as much as possible. Whenever we found parts of ES5 English prose to be ambiguous, we checked the browser implementations and were active on discussion groups [...]

Many of these efforts do not reach widespread adoption/acceptance, because they are not coupled well with the ongoing standardization efforts. I would to ask for your opinion on the matter and whether a formal semantics would be attractive to you. Thanks.

P.S.: I know the common criticism that a formal semantics does not get you anywhere since it is hard to verify correctness of compilation phases that then result in wasm code in the first place. However, CakeML for example is verified and it would be interesting to see this verification extend down to wasm code. Also, there might be more verified compilers that target wasm available in the future.

**Comment by rossberg** [comment]
Have you checked out the actual [language specification](https://webassembly.github.io/spec/core/index.html) for Wasm? It is currently under review as the W3C standard, and not only contains a formal semantics of the complete language, this semantics even represents the normative core of the spec. It is based on the formalisation from our [PLDI 2017 paper](https://github.com/WebAssembly/spec/blob/master/papers/pldi2017.pdf). I am confident to say that the Wasm spec is even more rigorous and precise than the SML Definition.

Moreover, this semantics has already been mechanised in Isabelle, see the [CPP 2018 paper](http://www.cl.cam.ac.uk/~caw77/papers/mechanising-and-verifying-the-webassembly-specification.pdf) by @conrad-watt (who also was previously involved in JsCert). That work also includes a verified interpreter. Two other mechanisations are currently under development, one in Coq by @swasey (who is also part of the RustBelt team), and another one using the K framework. ;)

**Comment by lorenzleutgeb** [comment]
Sweet! Except from exposing my poor search skills, this is a great and probably definitive answer :)

I saw the language specification, but did not find the mechanisation in Isabelle. For reference, the one mentioned can be downloaded [here](https://www.cl.cam.ac.uk/~caw77/).

**Comment by jwiegley** [comment]
@swasey Is it possible to look at and perhaps assist with your Coq development?

**Comment by jwiegley** [comment]
@swasey Any word on making this code publicly accessible?

**Comment by vzaliva** [comment]
@swasey I am also interested in WASM semantics in Coq.

**Comment by stefandeml** [comment]
@swasey is the Coq publicly available?

**Comment by sunfishcode** [comment]
See [this page](https://webassembly.org/news/2025-03-27-spectec/) for a more recent update on the state of mechanisation of the spec. If there are further questions, please file new issues!
